src/config/proxy.rs

```diff
@@ -0,0 +1,238 @@
+use http::uri::InvalidUri;
+use hyper_proxy::{Custom, Intercept, Proxy, ProxyConnector};
+use no_proxy::NoProxy;
+
+fn from_env(key: &str) -> Option<String> {
+    // use lowercase first and the uppercase
+    std::env::var(key.to_string().to_lowercase())
+        .ok()
+        .or_else(|| std::env::var(key.to_uppercase()).ok())
+}
+
+#[derive(serde::Deserialize, serde::Serialize, Clone, Default, Debug, PartialEq, Eq)]
+pub struct NoProxyInterceptor(NoProxy);
+
+impl NoProxyInterceptor {
+    fn intercept(self, expected_scheme: &'static str) -> Intercept {
+        Intercept::Custom(Custom::from(
+            move |scheme: Option<&str>, host: Option<&str>, port: Option<u16>| {
+                if scheme != Some(expected_scheme) {
+                    return false;
+                }
+                let matches = if let Some(host) = host {
+                    if let Some(port) = port {
+                        let url = format!("{}:{}", host, port);
+                        self.0.matches(&url) || self.0.matches(&host)
+                    } else {
+                        self.0.matches(&host)
+                    }
+                } else {
+                    false
+                };
```

Comment:
This could be done with adapters, but I'm on the fence if it improves readability. I think we *should* be doing the host match first before constructing the URL to compare it in any case.
```suggestion
                let matches = host.map(|host| {
                    self.0.matches(&host) || port.map(|port| {
                        let url = format!("{}:{}", host, port);
                        self.0.matches(&url)
                    }).unwrap_or(false)
                })
                .unwrap_or(false);
```